crates/pypi-types/src/metadata.rs

```diff
@@ -124,12 +128,78 @@ impl Metadata21 {
             provides_extras,
         })
     }
-}
 
-impl FromStr for Metadata21 {
-    type Err = Error;
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Self::parse(s.as_bytes())
+    /// Read the [`Metadata21`] from a source distribution's `PKG-INFO` file, if it uses Metadata 2.2
+    /// or later _and_ none of the required fields (`Requires-Python`, `Requires-Dist`, and
+    /// `Provides-Extra`) are marked as dynamic.
+    pub fn parse_pkg_info(content: &[u8]) -> Result<Self, Error> {
+        let headers = Headers::parse(content)?;
+
+        // To rely on a source distribution's `PKG-INFO` file, the `Metadata-Version` field must be
+        // present and set to a value of at least `2.2`.
+        let metadata_version = headers
+            .get_first_value("Metadata-Version")
+            .ok_or(Error::FieldNotFound("Metadata-Version"))?;
+        if metadata_version != "2.2" && metadata_version != "2.3" {
```

Comment:
We need to split, parse and check for >=2.2,<3, otherwise uv breaks when there's a metadata update